README.md
# Youtube time notes

This is a youtube notetaking app. You can take notes while watching video. When typing, the video will stop and wait till you finish your note. It also saves the timestamp of the video, when you started typing.

Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][Keep a Changelog] and this project adheres to [Semantic Versioning][Semantic Versioning].

## [Unreleased]

## Added

- option to insert video link into markdown output
- command capital `S` for save

## Fixed

- mobile keyboard enter handling

---

## [0.1.3] - 2021-09-13

Added new features and fixed some bugs.

### Added

- option to change video playback speed
- option to edit and delete individual notes
- button to start taking notes again from the note view
- about page
- export options in formats: txt, json, md
- code cleanup and documentation comments

### Changed

- how the notes are displayed

### Fixed

- added 2 newlines at the end of the note in export, so there is newline between them
- added trimming of the notes, so no redundant newline at the end and output will be consistent

### Removed

- unused buttons to change quality (not supported by youtube api) and show controls and commands, used about page instead

---

## [0.1.2] - 2021-09-11

Added new features and changed some behavior. Also changed the format of saved notes. No migration support available from the old one to the new one.

New format:

```JSON
{
    "id": "video-id",
    "title": "video-title",
    "notes": {
        "timestamp": "note text"
    }
}
```

### Added

- Styling with bootstrap
- Clicking on timestamp seeks video to given time, change cursor on hover to pointer
- Editing of notes on notetaking page
- Notes deletion
- Save command (s)
- Multiline note support (Shift+Enter)
- Left/Right arrow to seek in video

### Changed

- Unified width of all elements
- Better URL param parsing
- Safer element generation in list view
- Load existing notes if available
- Format of saved notes (contain id, title, and object of notes, where keys are positions in video in seconds)

### Deprecated

- Old format of saved notes (just an object of notes without any other metadata)

## [0.1.1] - 2020-05-01

### Fixed

- wrong version in changelog file

## [0.1.0] - 2020-05-01

Initial version

### Added

- Basic functionality for note taking
- Exporting notes in json format
- Saving notes in browser local storage

---

<!-- Links -->
[Keep a Changelog]: https://keepachangelog.com/
[Semantic Versioning]: https://semver.org/

<!-- Versions -->
[Unreleased]: https://github.com/makovako/youtube-time-notes/compare/v0.1.3...HEAD
[Released]: https://github.com/makovako/youtube-time-notes/releases
[0.1.3]: https://github.com/makovako/youtube-time-notes/releases/v0.1.3
[0.1.2]: https://github.com/makovako/youtube-time-notes/releases/v0.1.2
[0.1.1]: https://github.com/makovako/youtube-time-notes/releases/v0.1.1
[0.1.0]: https://github.com/makovako/youtube-time-notes/releases/v0.1.0


js/list.js
const main = document.querySelector("main");

let notes = [];

const list_group = document.createElement('div')
list_group.classList.add("list-group")

for (let i = 0; i < localStorage.length; i++) {
    note = JSON.parse(localStorage.getItem(localStorage.key(i)))
    notes.push(
        note
    );
}

notes.forEach((note) => {
    const div = document.createElement('div')
    div.classList.add("list-group-item")
    div.classList.add("list-group-item-action")
    div.classList.add("d-flex")
    div.classList.add("justify-content-between")
    div.classList.add("align-items-center")
    const a = document.createElement('a')
    const {id, title} = note
    a.href = `video.html?videoId=${id}`
    a.textContent = `${title} (${id})`
    div.appendChild(a)
    const btn = document.createElement('button')
    btn.classList.add("btn")
    btn.classList.add("btn-danger")
    btn.textContent = "Delete"
    btn.addEventListener('click', e => {
        e.preventDefault()
        localStorage.removeItem(id)
        location.reload()
    })
    div.appendChild(btn)
    list_group.append(div)
});

main.append(list_group)


js/video.js
const main = document.querySelector("main");
const videoId = new URLSearchParams(window.location.search).get("videoId");

const notes = localStorage.getItem(videoId);
const list_group = document.createElement('ul')
list_group.classList.add('list-group-flush', 'list-group')

const edit_note_form = document.getElementById('edit-note-form')
const delete_note_form = document.getElementById('delete-note-form')

const edit_note_input_area = document.getElementById('edit-note-input-area')

const start_notetaking_button = document.getElementById('start-notetaking-btn')

const data = JSON.parse(notes)

for (let [key, value] of Object.entries(data.notes)) {
    const li = document.createElement('li')
    li.classList.add('list-group-item', 'd-flex', 'justify-content-between', 'align-items-center')

    const div1 = document.createElement('div')
    div1.classList.add('d-flex', 'align-items-center')
    const div2 = document.createElement('div')
    div2.classList.add('d-flex', 'align-items-center')

    const timestamp = document.createElement('p')
    timestamp.classList.add('m-1')
    timestamp.textContent = format_current_time(key)
    const note = document.createElement('p')
    note.classList.add('m-1')
    note.setAttribute('style', 'white-space: pre-wrap; overflow-wrap: break-word;');
    note.textContent = value

    const edit_btn = document.createElement('button')
    edit_btn.classList.add('btn', 'btn-warning', 'm-1')
    edit_btn.type = "button"
    edit_btn.dataset.toggle = "modal"
    edit_btn.dataset.target = "#editNoteModal"
    edit_btn.textContent = "Edit"
    edit_btn.addEventListener('click', () => {
        // Remember, which note to edit
        edit_note_form.dataset.note_id = key
        edit_note_input_area.value = value
    })

    const delete_btn = document.createElement('button')
    delete_btn.classList.add('btn', 'btn-danger', 'm-1')
    delete_btn.type = "button"
    delete_btn.dataset.toggle = "modal"
    delete_btn.dataset.target = "#deleteNoteModal"
    delete_btn.textContent = "Delete"
    delete_btn.addEventListener('click', () => {
        // Remember, which note to delete
        delete_note_form.dataset.note_id = key
    })

    div1.appendChild(timestamp)
    div1.appendChild(note)
    div2.appendChild(edit_btn)
    div2.appendChild(delete_btn)
    li.appendChild(div1)
    li.appendChild(div2)
    list_group.append(li)
}

main.append(list_group)

// Note editing handling

edit_note_form.addEventListener('submit', (e) => {
    e.preventDefault()
    key = e.target.dataset.note_id
    data.notes[key] = edit_note_input_area.value.trim()
    localStorage.setItem(videoId, JSON.stringify(data));
    location.reload()
})

delete_note_form.addEventListener('submit', (e) => {
    e.preventDefault()
    key = e.target.dataset.note_id
    delete data.notes[key]
    localStorage.setItem(videoId, JSON.stringify(data));
    location.reload()
})

start_notetaking_button.addEventListener('click', () => {
    window.location.href = `notetaking.html?videoId=${videoId}`
})

// Note export handling

const export_md_form = document.getElementById('export-md-form')
const export_text_form = document.getElementById('export-text-form')
const export_json_form = document.getElementById('export-json-form')

const filename_md = document.getElementById('filename-md')
const no_timestamps_md = document.getElementById('no-timestamps-md')
const timestamps_text_md = document.getElementById('timestamps-text-md')
const timestamps_list_md = document.getElementById('timestamps-list-md')
const timestamps_headings_md = document.getElementById('timestamps-headings-md')
const md_as_link = document.getElementById('md-as-link')
const md_insert_link = document.getElementById('md-insert-link')

const filename_text = document.getElementById('filename-text')
const insert_link_text = document.getElementById('insert-link-text')

const filename_json = document.getElementById('filename-json')

no_timestamps_md.addEventListener('change', (e) => {
    md_as_link.checked = false
    md_as_link.disabled = true
});

[timestamps_text_md, timestamps_list_md, timestamps_headings_md]
.forEach(elm => {
    elm.addEventListener('change', (e) => {
        md_as_link.disabled = false
    })
})

export_md_form.addEventListener('submit', (e) => {
    e.preventDefault()
    let file_content = ""
    for (let [key, value] of Object.entries(data.notes)) {
        let timestamp = ""
        if (md_as_link.checked) {
            link = `https://www.youtube.com/watch?v=${videoId}&t=${key}`
            timestamp = `[${format_current_time(key)}](${link})`
        } else {
            timestamp = format_current_time(key)
        }
        if (timestamps_text_md.checked) {
            file_content += `${timestamp}\n\n`
        }
        if (timestamps_list_md.checked) {
            file_content += `- ${timestamp}\n\n`
        }
        if (timestamps_headings_md.checked) {
            file_content += `# ${timestamp}\n\n`
        }
        file_content += value + "\n\n"
    }
    if (md_insert_link.checked) {
        file_content += `[source](https://www.youtube.com/watch?v=${videoId})\n\n`
    }
    const dataStr = "data:text/plain;charset=utf-8," + encodeURIComponent(file_content)
    const downloadAnchorNode = document.createElement('a');
    let filename = filename_md.value
    if (filename.length === 0) {
        filename = videoId
    }
    if (!filename.endsWith(".md")) {
        filename += ".md"
    }
    downloadAnchorNode.setAttribute("href", dataStr);
    downloadAnchorNode.setAttribute("download", filename);
    downloadAnchorNode.click();
    downloadAnchorNode.remove();
})

export_text_form.addEventListener('submit', (e) => {
    e.preventDefault()
    let file_content = ""
    for (let [key, value] of Object.entries(data.notes)) {
        if (insert_link_text.checked) {
            file_content += `${format_current_time(key)}\n\n`
        }
        file_content += value + "\n\n"
    }

    const dataStr = "data:text/plain;charset=utf-8," + encodeURIComponent(file_content)
    const downloadAnchorNode = document.createElement('a');
    let filename = filename_text.value
    if (filename.length === 0) {
        filename = videoId
    }
    if (!filename.endsWith(".txt")) {
        filename += ".txt"
    }
    downloadAnchorNode.setAttribute("href", dataStr);
    downloadAnchorNode.setAttribute("download", filename);
    downloadAnchorNode.click();
    downloadAnchorNode.remove();
})

export_json_form.addEventListener('submit', (e) => {
    e.preventDefault()
    const dataStr = "data:text/json;charset=utf-8," + encodeURIComponent(notes)
    const downloadAnchorNode = document.createElement('a');
    let filename = filename_json.value
    if (filename.length === 0) {
        filename = videoId
    }
    if (!filename.endsWith(".json")) {
        filename += ".json"
    }
    downloadAnchorNode.setAttribute("href", dataStr);
    downloadAnchorNode.setAttribute("download", filename);
    downloadAnchorNode.click();
    downloadAnchorNode.remove();
})

// Helper functions

function format_current_time(current_time) {
    let hrs = Math.floor(current_time / 3600).toString();
    let mins = Math.floor((current_time % 3600) / 60).toString();
    let secs = Math.floor(~~current_time % 60).toString();
    return `${hrs.padStart(2,"0")}:${mins.padStart(2,"0")}:${secs.padStart(2,"0")}`;
}

js/notetaking.js
// Video param
const videoId = new URLSearchParams(window.location.search).get("videoId");

// Elements

const input_area = document.getElementById("input-area");
input_area.value = ""

const notes = document.getElementById("notes");
const save = document.getElementById("save");
const speed_body = document.getElementById("pick-speed-modal-body")

// Loading youtube script
var tag = document.createElement("script");

tag.src = "https://www.youtube.com/iframe_api";
var firstScriptTag = document.getElementsByTagName("script")[0];
firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);

// Computing properties

const container = document.querySelector('.container')
const container_padding_left = window.getComputedStyle(container, null).getPropertyValue('padding-left').slice(0,-2)
const container_padding_right = window.getComputedStyle(container, null).getPropertyValue('padding-right').slice(0,-2)

const video_width = container.clientWidth - container_padding_left - container_padding_right
const video_height = Math.floor(video_width * 9 / 16)

// Creating iframe
let player;
function onYouTubeIframeAPIReady() {
    player = new YT.Player("player", {
        height: video_height,
        width: video_width,
        videoId: videoId,
        playerVars: {
            //   controls: 0,
            //   autoplay: 1,
            origin: "http://127.0.0.1:8080",
            //   disablekb: 1,
        },
        events: {
            onReady: onPlayerReady,
            onStateChange: onPlayerStateChange,
        },
    });
}

// player.addEventListener('onReady', event => {
//     event.target.playVideo();
//     player.playVideo()
//     console.log('here');

// })

function format_current_time(current_time) {
    let hrs = Math.floor(current_time / 3600).toString();
    let mins = Math.floor((current_time % 3600) / 60).toString();
    let secs = Math.floor(~~current_time % 60).toString();
    return `${hrs.padStart(2,"0")}:${mins.padStart(2,"0")}:${secs.padStart(2,"0")}`;
}

// Load existing data

let data = JSON.parse(localStorage.getItem(videoId))

if (data) {
    for (let [key, value] of Object.entries(data.notes)) {
        li = createNoteLi(key, value)
        notes.append(li)
    }
} else {
    data = {}
}

// Create custom commands (like enter to play)
const commands = [
    {
        command: "p\n",
        description: "play",
        run: () => player.playVideo(),
    },
    {
        command: "\n",
        description: "play",
        run: () => player.playVideo(),
    },
    {
        command: "s\n",
        description: "save",
        run: () => {
            localStorage.setItem(videoId, JSON.stringify(data));
            window.location.href = "list.html";
        }
    },
    {
        command: "S\n",
        description: "save",
        run: () => {
            localStorage.setItem(videoId, JSON.stringify(data));
            window.location.href = "list.html";
        }
    }
];

// Notetaking logic

input_area.addEventListener("keyup", (e) => {
    if (e.key === "Enter") {
        if (e.shiftKey) {
            // allow new line in note with shift+enter
            return
        }
        // Run custom command
        const cmd = commands.filter((command) => input_area.value === command.command);
        if (cmd.length > 0) {
            cmd[0].run()
            input_area.value = ""
            return
        }

        // current time works as an id for note
        let currTime = player.getCurrentTime()
        currTime = Math.floor(currTime)

        note_text = input_area.value
        note_text = note_text.trim()
        li = createNoteLi(currTime, note_text)
        notes.append(li)
        data.notes[currTime] = note_text;
        input_area.value = "";
        player.playVideo();
    } else if (player.getPlayerState() === 1) {
        let currTime = player.getCurrentTime()
        // Allow user to fast-forward/rewind with arrows
        if (e.key === "ArrowRight") {
            player.seekTo(currTime + 10, true)
        }else if (e.key === "ArrowLeft") {
            player.seekTo(currTime - 10, true)
        } else {
            // Any other key means pause video and start typing note
            player.pauseVideo();
        }
    }
});

// Buttons actions

save.addEventListener("click", () => {
    localStorage.setItem(videoId, JSON.stringify(data));
    window.location.href = "list.html";
});

// Player events

function onPlayerReady(event) {
    const speeds = event.target.getAvailablePlaybackRates()
    speeds.forEach((speed) => {
        const btn = document.createElement('button')
        btn.classList.add('btn', 'btn-outline-warning', 'm-1')
        btn.textContent = speed
        btn.addEventListener('click', (event) => {
            event.preventDefault()
            player.setPlaybackRate(speed)
        })
        // Close modal when speed chosen
        btn.dataset.dismiss = "modal"
        speed_body.appendChild(btn)
    })

    const {video_id, title} = event.target.getVideoData()
    data["id"] = video_id
    data["title"] = title
    if (!data.notes) {
        data["notes"] = {}
    }
    input_area.focus()
}

function onPlayerStateChange(event) {}

// Helper functions

function createNoteLi(currTime, noteText) {
    // Reuse existing note if possible (edit note)
    let li = document.getElementById(currTime)
    if (!li) {
        li = document.createElement("li");
        li.id = currTime
        li.classList.add('list-group-item', 'd-flex', 'align-items-center')
    }

    // Clear note content
    while (li.firstChild) {
        li.removeChild(li.lastChild);
    }

    // Create clickable timestamp
    const timestamp = document.createElement('a')
    timestamp.href = "#"
    timestamp.textContent = format_current_time(currTime)
    timestamp.classList.add('m-1', 'mr-2', 'p-1')
    timestamp.title = `Seek video to time ${format_current_time(currTime)} and edit note`
    timestamp.addEventListener('click', e => {
        e.preventDefault()
        player.pauseVideo()
        player.seekTo(currTime, true)
        player.pauseVideo()
        input_area.value = data.notes[currTime]
        input_area.focus()
    })

    const note = document.createElement('p')
    note.setAttribute('style', 'white-space: pre-wrap; overflow-wrap: break-word;');
    note.classList.add('m-1')
    note.textContent = noteText

    li.appendChild(timestamp)
    li.appendChild(note)
    return li
}
